### DISCUSSION TITLE
How to disable mangling of variable names during transform()?

### DISCUSSION THREAD

**Discussion by gogo9th** [babel/babel#12448] [Q&A]
Hi,

When I transform code by using Babel, the original variable names are mangled, which causes troubles in my application scenario. So my goal is to disable Babel's variable name mangling. To do this, the following is my Babel transformation program:

```
// transform.js`
if (args.length != 1)
{  console.log("Usage: node transform <target .js>")
   process.exit(0);
}
const code = fs.readFileSync(args[0], 'utf8');
var result = babel.transform(code,
      {  retainLines: true,
         compact: false,
         sourceType: 'unambiguous',
         sourceMaps: true,
         presets: ['@babel/preset-env'],
         plugins: [
            ["transform-es2015-modules-commonjs-simple",
               { 'noMangle': true}
            ]
         ]
      }).code;
console.log(result)
```

And the following is my input program:

```
// problem.js`
var s_d = function() {
   var s_b = 1;
   s_d = 1;
}
```

I transform the code by doing the following:

`node transform problem.js`

And my output is the following:

```
"use strict";var _s_d = function s_d() {
  var s_b = 1;
  _s_d = 1;
};
```

The variable `s_d` is changed to `_s_d`. How can I disable mangling? I would appreciate any clue. Thanks.

**Comment by gogo9th** [comment]
If I comment out `presets: ['@babel/preset-env']`, then it works. But what's the disadvantage or potential problem if I comment out presets? And is there a way to keep both presets variable names?

**Comment by gogo9th** [comment]
Actually, the `"transform-es2015-modules-commonjs-simple"` plugin seems buggy. It shadows the `"@babel/plugin-transform-modules-commonjs"` plugin's `{ 'strictMode': false }` option. Also, it doesn't transform the following simple code correctly:


`this;`


. . . because the code gets transforms to:

`"use strict";undefined;`

So we need another solution..

**Comment by gogo9th** [comment]
This works:

```
const code = fs.readFileSync(args[0], 'utf8');
var result = babel.transform(code,
      {  retainLines: true,
         compact: false,
         sourceType: 'unambiguous',
         sourceMaps: true,
         //presets: ['@babel/preset-env'],
         plugins: [
            "@babel/transform-runtime",
            /*[ "@babel/plugin-transform-modules-commonjs",
               { 'strictMode': false }
            ],*/
            ["transform-es2015-modules-commonjs-simple",
               { 'noMangle': true,
                'allowTopLevelThis' : true,
                'strictMode': false
               },
            ]
         ]
      }).code;
```

**Comment by nicolo-ribaudo** [comment]
The plugin that is renaming that variable is `transform-function-name`: you can pass `{ exclude: ["babeltransform-function-name"] }` as the options of `@babel/preset-env` and it should work.

Btw, what is your goal with Babel?
1. Only transform modules to commonjs -> Don't use `@babel/preset-env`
2. Transpile all the modern JS features so that they work in older engines -> Use `@babel/preset-env`, maybe setting the `targets` option.

**Comment by gogo9th** [comment]
#2 is the goal.

I included the following, but doesn't seem to work:
```
var result = babel.transform(code,
      {  retainLines: true,
         filename: 'text.js',
         compact: false,
         sourceType: 'unambiguous',
         sourceMaps: true,
         presets: [  '@babel/preset-env', 
                     {"exclude" : ["plugin-transform-function-name"] }         
                  ]
      }).code;


```

**Comment by gogo9th** [comment]
If I manually comment out the following two lines of `node_modules/@babel/plugin-transform-function-name/lib/index.js`, it works:

```
    name: "transform-function-name",
    visitor: {
      FunctionExpression: {
        exit(path) {
          if (path.key !== "value" && !path.parentPath.isObjectProperty()) {
            //const replacement = (0, _helperFunctionName.default)(path);
            //if (replacement) path.replaceWith(replacement);
          }
        }
      },
```
But this is not a clean solution... Is there a decent way to disable `transform-function-name` without hacking the module's source code?

**Reply by nicolo-ribaudo** [reply]
The last config you posted should already disabled it. Try passing `debug: true` to `@babel/preset-env` to see what plugin it's enabling.
Also, do you have another Babel config file? Maybe it's causing the inclusion of the plugin even if you excluded it from `babel.transform`'s options.

**Reply by gogo9th** [reply]
Hi, I tried to give the debug option, but it does not work:

```
var result = babel.transform(code,
      {  retainLines: true,
         filename: '',
         compact: false, 
         sourceType: 'unambiguous',
         sourceMaps: true,
         presets: [  '@babel/preset-env', 
                     //{"exclude" : ["@babel/plugin-transform-function-name"] }, 
                     {"debug" : true },
                  ], 
      }).code;
```

/home/skyer/Desktop/expose_runner/node_modules/@babel/core/lib/config/full.js:192
      throw e;
      ^

```
Error: [BABEL] /home/skyer/Desktop/expose_runner/chrome-experiment-result: Unknown option: .debug. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.
- Maybe you meant to use
"preset": [
  ["@babel/preset-env", {
  "debug": true
}]
]
To be a valid preset, its name and options should be wrapped in a pair of brackets
```
